src/providers-registry/providers-registry-contract.ts

```diff
@@ -0,0 +1,270 @@
+    if (newProvider.manifests !== undefined) {
+      throw new ContractError("Manifest should be added add with separate add-provider-manifest function.");
+    }
+
+    // initializing adminsPools array if not defined
+    if (newProvider.adminsPool === undefined) {
+      data.provider.adminsPool = [];
+    }
+
+    // adding caller as default admin if not found in adminsPool
+    if (!newProvider.adminsPool.includes(caller)) {
+      newProvider.adminsPool.push(caller);
+    }
+
+    newProvider.manifests = [];
+    newProvider.registerHeight = blockHeight;
+    newProvider.profile.id = `provider_${SmartWeave.transaction.id}`;
+
+    providers[caller] = newProvider;
+
+    trace("STATE", state);
+
+    return {state};
+  }
+
+  if (input.function === "remove-provider") {
+    const data = input.data as RemoveProviderData;
+    checkProviderId(data.providerId);
+    checkProviderExists(data.providerId);
+    checkPrivileges(caller, data.providerId);
+
+    delete providers[data.providerId];
+
+    trace("STATE", state);
+
+    return {state};
+  }
+
+  if (input.function === "add-provider-manifest") {
+    const addManifestData = input.data as AddProviderManifestData;
+    const manifestProviderId = addManifestData.providerId;
+    checkProviderId(manifestProviderId);
+    checkProviderExists(manifestProviderId);
+    checkPrivileges(caller, manifestProviderId);
+
+    const manifestData = addManifestData.manifestData;
+
+    if (manifestData === undefined) {
+      throw new ContractError("Manifest data not set.");
+    }
+
+    if (manifestData.manifest === undefined) {
+      throw new ContractError("Manifest not set.");
+    }
+
+    if (manifestData.changeMessage === undefined || manifestData.changeMessage.length === 0) {
+      throw new ContractError("Change message is not set.");
+    }
+
+    // TODO: manifest properties validation
+
+    // note: this is safe, as manifests array is initialized when provider is added
+    providers[manifestProviderId].manifests.push(
+      {
+        uploadBlockHeight: blockHeight,
+        manifest: manifestData.manifest,
+        changeMessage: manifestData.changeMessage,
+        lockedHours: addManifestData.lockedHours || 0
+      }
+    );
+
+    trace("STATE", state);
+
+    return {state};
+  }
+
+  if (input.function === "add-provider-admin") {
+    const data = input.data as AddProviderAdminData;
+    checkProviderId(data.providerId);
+    checkProviderExists(data.providerId);
+    checkPrivileges(caller, data.providerId);
+
+    // note: this is safe, as adminsPool array is initialized when provider is added
+    providers[data.providerId].adminsPool.push(...data.admins);
+
+    trace("STATE", state);
+
+    return {state};
+  }
+
+  // admin-only: switch trace state
+  if (input.function === "switch-trace") {
+    checkIsContractAdmin(caller);
+    state.trace = !state.trace;
+
+    trace("END STATE", state);
+
+    return {state}
+  }
+
+  if (input.function === "add-contract-admins") {
```

Comment:
The same comment about validation of admin address.
Should we add similar functions for revoking admin privileges?